### DISCUSSION TITLE
0.35.0 refactor of root_path handling is potentially returning incorrect route

### DISCUSSION THREAD

**Discussion by gabriel-f-santos** [encode/starlette#2601] [Q&A]
### Discussed in https://github.com/encode/starlette/discussions/2495

<div type='discussions-op-text'>

<sup>Originally posted by **robbielaw** February  8, 2024</sup>

When running starlette/fastapi behing a proxy we got some errors when calling the app directly.

when calling curl http://127.0.0.1:8000/products/, we get:
root_path: _/product_
scope['path']: _/products_
route_path: _s_ (after apply the regex - https://github.com/encode/starlette/blob/master/starlette/_utils.py#L96)
because the the root_path to be removed if calling directly the app is **"product"** and endpoint path is **/products**

**Example Scenario**:
libs: 
```
starlette==0.37.2
uvicorn==0.29.0
```
gist: https://gist.github.com/gabriel-f-santos/df75213fe6860e369a1e096611a767e5
run: uvicorn main:app --root-path /product

Code causing the error:
https://github.com/encode/starlette/blob/master/starlette/_utils.py#L96

```
def get_route_path(scope: Scope) -> str:
    root_path = scope.get("root_path", "")
    route_path = re.sub(r"^" + root_path, "", scope["path"])
    return route_path
```

Fix suggestion:

```
def get_route_path(scope: Scope) -> str:
    root_path = scope.get("root_path", "")
    route_path = re.sub(r"^" + root_path + r"(?=/|$)", "", scope["path"])
    return route_path
```

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/starlette/issues/2599">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/starlette/issues/2599/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/starlette/issues/2599/pledge.svg">
</picture>
</a>

**Comment by amol-** [comment]
Based on my understanding of https://github.com/django/asgiref/issues/229#issuecomment-765583185 @gabriel-f-santos raises a valid point. The comment states that the expectation when the ASGI specification was written was that web frameworks would strip the `root_path` themselves when performing routing.

This is what should be happening in https://github.com/encode/starlette/blob/master/starlette/_utils.py#L97-L98 (even though the check for `"/"` is confusing without a comment, as `startswith` was already checked previously)

**Reply by Kludex** [reply]
Gabriel's PR was merged...

**Reply by amol-** [reply]
Thanks for confirming, I just wanted to add a reference to the asgi specs discussion to confirm it's the intended behavior.

Also referencing here https://github.com/encode/uvicorn/discussions/2490#discussion-7349238 which causes unexpected duplication of root_path even if starlette removes it

**Reply by Kludex** [reply]
It's a bit hard to follow the `root_path` discussions... And the topic was discussed many times... If you are bringing a new point that was not covered before (which is hard to believe given the number of times this was revisited), it would be more useful to provide code examples, and comparison with other frameworks.

**Reply by amol-** [reply]
@Kludex , my comment and link to uvcorn is an attempt to make it more clear the behaviours (expected and implemented) to anyone finding this issue. As you mentioned it's a bit confusing the situation with `root_path` and the same questions have been brought up many times. I thought about adding the information as it would have been helpful for me when I was researching the issue

**Reply by Kludex** [reply]
Ah ok! Thanks!
